### PR TITLE
Add option to copy repo from remote

### PR DESCRIPTION
**Pull request by patriotyk**

### CODE REVIEW

File: tools/main.cc [hunk 1]
```diff
@@ -135,15 +143,22 @@ int main(int argc, char **argv) {
     assert(0);
   }
 
-  OSTreeRepo repo(repo_path);
+  TreehubServer pull_server;
+  OSTreeRepo *repo;
```
Comment (cajun-rat) [inline tools/main.cc]: What about using shared_ptr here?

File: tools/ostree_http_repo.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                  << (root_ / "config").string();
+      return false;
+    } catch (const pt::ptree_error &error) {
+      LOG_WARNING << "Could not find core.mode in OSTree config file";
+      return false;
+    }
+  } else {
+    return false;
+  }
+  return true;
```
Comment (cajun-rat) [inline tools/ostree_http_repo.cc]: This line is unreachable: can you remove it, please?

File: tools/ostree_http_repo.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#include "ostree_object.h"
+#include "ostree_ref.h"
+#include "treehub_server.h"
+#include "ostree_repo.h"
+
+
+class OSTreeHttpRepo : public OSTreeRepo {
+ public:
+  explicit OSTreeHttpRepo(const TreehubServer &server) : server_(server), root_(std::string("/tmp/")+boost::filesystem::unique_path().string())
+ { boost::filesystem::create_directories(root_); }
```
Comment (cajun-rat) [inline tools/ostree_http_repo.h]: What about deleting this temporary directory (and its contents) in the dtor?

File: tools/ostree_ref.cc [hunk 4]
```diff
@@ -14,8 +14,43 @@ using std::stringstream;
...
+  CURL *curl_handle = curl_easy_init();
+  serve_repo.InjectIntoCurl(Url(), curl_handle);
+  curl_easy_setopt(curl_handle, CURLOPT_WRITEFUNCTION,
+                   &OSTreeRef::curl_handle_write);
+  curl_easy_setopt(curl_handle, CURLOPT_WRITEDATA, this);
+  curl_easy_setopt(curl_handle, CURLOPT_VERBOSE, get_curlopt_verbose());
+  curl_easy_setopt(curl_handle, CURLOPT_FAILONERROR, true);
+  CURLcode rc = curl_easy_perform(curl_handle);
+  if (rc != CURLE_OK) {
+    is_valid = false;
```
Comment (cajun-rat) [inline tools/ostree_ref.cc]: What is the code path that sets is_valid to true for fetch-over-http case?

**Comment by cajun-rat** [comment]
Overall I like the design. With a couple of changes, this should be good to merge.

File: tools/ostree_ref.cc (hunk 4 above)
Comment (patriotyk) [reply tools/ostree_ref.cc]: Oh really good catch. It was working accidentally

File: tools/ostree_http_repo.h (hunk 3 above)
Comment (patriotyk) [reply tools/ostree_http_repo.h]: +

File: tools/ostree_http_repo.cc (hunk 2 above)
Comment (patriotyk) [reply tools/ostree_http_repo.cc]: +

File: tools/main.cc (hunk 1 above)
Comment (patriotyk) [reply tools/main.cc]: +

**Comment by patriotyk** [comment]
Updated

**Comment by cajun-rat** [comment]
Sweet, this works a charm :)

**Pull request merged**